### PR TITLE
Add `Toolbar` implementation for HTML renderer

### LINKED ISSUE #316: Add support for `toolbar` modifier and related types
Relevant APIs:

```swift
public protocol ToolbarContent {
  associatedtype Body : ToolbarContent
  @ToolbarContentBuilder var body: Self.Body { get }
}

@_functionBuilder public struct ToolbarContentBuilder {
  public static func buildBlock<Content>(
    _ content: Content
  ) -> some ToolbarContent where Content : ToolbarContent
  
  public static func buildBlock<Content>(
    _ content: Content
  ) -> some CustomizableToolbarContent where Content : CustomizableToolbarContent
}

extension View {
  public func toolbar<Content>(
    @ViewBuilder content: () -> Content
  ) -> some View where Content : View
  
  public func toolbar<Content>(
    @ToolbarContentBuilder content: () -> Content
  ) -> some View where Content : ToolbarContent
  
  public func toolbar<Content>(
    id: Swift.String,
    @ToolbarContentBuilder content: () -> Content
  ) -> some View where Content : CustomizableToolbarContent
}

public struct ToolbarItem<ID, Content> : ToolbarContent where Content : View {
  public typealias Body = Never
}

extension ToolbarItem where ID == Void {
  public init(
    placement: ToolbarItemPlacement = .automatic,
    @ViewBuilder content: () -> Content
  )
}

extension ToolbarItem : CustomizableToolbarContent where ID == String {
  public init(
    id: String,
    placement: ToolbarItemPlacement = .automatic,
    showsByDefault: Swift.Bool = true,
    @ViewBuilder content: () -> Content
  )
}

extension ToolbarItem : Identifiable where ID : Hashable {
  public var id: ID { get }
}

public struct ToolbarItemGroup<Content> : ToolbarContent where Content : View {
  public init(
    placement: ToolbarItemPlacement = .automatic,
    @ViewBuilder content: () -> Content
  )
  public typealias Body = Never
}

public struct ToolbarItemPlacement {
  public static let automatic: Self
  public static let primaryAction: Self
  public static let confirmationAction: Self
  public static let cancellationAction: Self
  public static let destructiveAction: Self
  public static let navigationBarLeading: Self
  public static let navigationBarTrailing: Self
  public static let bottomBar: Self
}

extension View {
  public func toolbar<Content>(
    @ViewBuilder content: () -> Content
  ) -> some View where Content : View
  
  public func toolbar<Content>(
    @ToolbarContentBuilder content: () -> Content
  ) -> some View where Content : ToolbarContent
  
  public func toolbar<Content>(
    id: Swift.String,
    @ToolbarContentBuilder content: () -> Content
  ) -> some View where Content : CustomizableToolbarContent
}
```

### PR DESCRIPTION
**Pull request by carson-katri**
`Toolbar` is new in SwiftUI. It is coupled fairly closely with `NavigationView`, so this should be integrated with that somehow (#130). I made it similar to macOS which allows more than a leading/trailing `ToolbarItem`.

Resolves #316.

### CODE REVIEW

File: Sources/TokamakDOM/Resources/TokamakStyles.swift
Comment (ie-ahm-robox) [inline Sources/TokamakDOM/Resources/TokamakStyles.swift:84]: - :no_entry_sign: Line should be 100 characters or less: currently 123 characters (`line_length`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Toolbar.swift#L115** - Line should be 100 characters or less: currently 123 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Toolbar.swift#L119** - Line should be 100 characters or less: currently 113 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Toolbar.swift#L124** - Line should be 100 characters or less: currently 121 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakDOM/Views/ScrollView.swift#L41** - Line should be 100 characters or less: currently 115 characters (`line_length`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against b97f658d4dbb22047916557c6137f129d2e93011
</p>

File: Sources/TokamakCore/Views/Toolbar.swift
Comment (ie-ahm-robox) [inline Sources/TokamakCore/Views/Toolbar.swift:124]: - :no_entry_sign: Line should be 100 characters or less: currently 121 characters (`line_length`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Toolbar.swift#L115** - Line should be 100 characters or less: currently 123 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Toolbar.swift#L119** - Line should be 100 characters or less: currently 113 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakDOM/Resources/TokamakStyles.swift#L84** - Line should be 100 characters or less: currently 123 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakDOM/Views/ScrollView.swift#L41** - Line should be 100 characters or less: currently 115 characters (`line_length`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against 75cb3960ffd5e4fc57226f5b722f214e00b76002
</p>

File: Sources/TokamakCore/Views/Toolbar.swift
Comment (ie-ahm-robox) [inline Sources/TokamakCore/Views/Toolbar.swift:115]: - :no_entry_sign: Line should be 100 characters or less: currently 123 characters (`line_length`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Toolbar.swift#L119** - Line should be 100 characters or less: currently 113 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Toolbar.swift#L124** - Line should be 100 characters or less: currently 121 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakDOM/Resources/TokamakStyles.swift#L84** - Line should be 100 characters or less: currently 123 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakDOM/Views/Layout/ScrollView.swift#L39** - Line should be 100 characters or less: currently 115 characters (`line_length`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against 4430bab69aaf2e2a8b0d166d074b70028828127e
</p>

File: Sources/TokamakStaticHTML/Resources/TokamakStyles.swift
Comment (ie-ahm-robox) [inline Sources/TokamakStaticHTML/Resources/TokamakStyles.swift:87]: - :no_entry_sign: Line should be 100 characters or less: currently 123 characters (`line_length`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L26** - SwiftLint rule 'line_length' did not trigger a violation in the disabled region. Please remove the disable command. (`superfluous_disable_command`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L256** - SwiftLint rule 'line_length' did not trigger a violation in the disabled region. Please remove the disable command. (`superfluous_disable_command`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L452** - File should contain 400 lines or less: currently contains 452 (`file_length`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against 2083dbb3c55bbd2f5724c8673ef90aa2e28e500c
</p>

File: Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift
Comment (ie-ahm-robox) [inline Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift:452]: - :no_entry_sign: File should contain 400 lines or less: currently contains 452 (`file_length`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L26** - SwiftLint rule 'line_length' did not trigger a violation in the disabled region. Please remove the disable command. (`superfluous_disable_command`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L256** - SwiftLint rule 'line_length' did not trigger a violation in the disabled region. Please remove the disable command. (`superfluous_disable_command`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakStaticHTML/Resources/TokamakStyles.swift#L88** - Line should be 100 characters or less: currently 123 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakStaticHTML/Views/Navigation/NavigationView.swift#L22** - Line should be 100 characters or less: currently 111 characters (`line_length`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against 01d7514723a8cb03f255f8e76da9c4ed867adaab
</p>

File: Sources/TokamakStaticHTML/Views/Navigation/NavigationView.swift
Comment (ie-ahm-robox) [inline Sources/TokamakStaticHTML/Views/Navigation/NavigationView.swift:22]: - :no_entry_sign: Line should be 100 characters or less: currently 111 characters (`line_length`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L26** - SwiftLint rule 'line_length' did not trigger a violation in the disabled region. Please remove the disable command. (`superfluous_disable_command`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L256** - SwiftLint rule 'line_length' did not trigger a violation in the disabled region. Please remove the disable command. (`superfluous_disable_command`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L452** - File should contain 400 lines or less: currently contains 452 (`file_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakStaticHTML/Resources/TokamakStyles.swift#L88** - Line should be 100 characters or less: currently 123 characters (`line_length`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against 459636fbca2fe1cccae656fc4fead587b3d51d40
</p>

File: Sources/TokamakCore/Views/Navigation/NavigationView.swift
Comment (ie-ahm-robox) [inline Sources/TokamakCore/Views/Navigation/NavigationView.swift:20]: - :no_entry_sign: Initializing an optional variable with nil is redundant. (`redundant_optional_initialization`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Preferences/PreferenceStore.swift#L68** - SwiftLint rule 'force_cast' did not trigger a violation in the disabled region. Please remove the disable command. (`superfluous_disable_command`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L26** - SwiftLint rule 'line_length' did not trigger a violation in the disabled region. Please remove the disable command. (`superfluous_disable_command`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L256** - SwiftLint rule 'line_length' did not trigger a violation in the disabled region. Please remove the disable command. (`superfluous_disable_command`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L452** - File should contain 400 lines or less: currently contains 452 (`file_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakStaticHTML/Resources/TokamakStyles.swift#L87** - Line should be 100 characters or less: currently 123 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakStaticHTML/Views/Navigation/NavigationView.swift#L21** - Line should be 100 characters or less: currently 111 characters (`line_length`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against 7ae483a61668bd6786aed0ce50d724f111c135de
</p>

File: Sources/TokamakCore/Preferences/PreferenceStore.swift
Comment (ie-ahm-robox) [inline Sources/TokamakCore/Preferences/PreferenceStore.swift:58]: - :no_entry_sign: SwiftLint rule 'force_cast' did not trigger a violation in the disabled region. Please remove the disable command. (`superfluous_disable_command`)

File: Sources/TokamakStaticHTML/Views/Navigation/NavigationView.swift
Comment (ie-ahm-robox) [inline Sources/TokamakStaticHTML/Views/Navigation/NavigationView.swift:21]: - :no_entry_sign: Line should be 100 characters or less: currently 107 characters (`line_length`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Preferences/PreferenceStore.swift#L24** - Line should be 100 characters or less: currently 114 characters (`line_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/NavigationView.swift#L20** - Initializing an optional variable with nil is redundant. (`redundant_optional_initialization`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L26** - SwiftLint rule 'line_length' did not trigger a violation in the disabled region. Please remove the disable command. (`superfluous_disable_command`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L256** - SwiftLint rule 'line_length' did not trigger a violation in the disabled region. Please remove the disable command. (`superfluous_disable_command`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Navigation/ToolbarContentBuilder.swift#L452** - File should contain 400 lines or less: currently contains 452 (`file_length`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakStaticHTML/Resources/TokamakStyles.swift#L87** - Line should be 100 characters or less: currently 123 characters (`line_length`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against 0a947b891d9f8285e4decc7e9bd25a861ffcec32
</p>

**Comment by carson-katri** [comment]
I think I have this working to a reasonable degree now. Here are some screenshots, most to show the different `ToolbarItemPlacement`s:

Screen Shot 2021-01-20 at 1 41 07 PM
Screen Shot 2021-01-20 at 1 42 57 PM

And here's the `Toolbar` in the demo:

Screen Shot 2021-01-20 at 2 07 15 PM

Try it out and let me know if the design of it should be tweaked at all.

**Comment by MaxDesiatov** [comment]
Do we need some spacing between "Some nav-text" and "Tokamak", or is that blocked by the absence of `spacing` on `HStack`?

**Comment by carson-katri** [comment]
I added some spacing. This isn't using a normal `HStack`, so #272 isn't blocking it.

Some limitations I noticed:

* Preferences can be slow, but I don't really know how we would speed them up...
* When the `navigationTitle` changes it doesn't re-render the toolbar because there is no State/ObservableObject change. I could possibly work around this by using `onPreferenceChange` that would update some state instead of using `_delay` and `_force` to access the preference values. But again, that would severely slow it down.

File: Sources/TokamakCore/Modifiers/Effects/OpacityEffect.swift
Comment (MaxDesiatov) [inline Sources/TokamakCore/Modifiers/Effects/OpacityEffect.swift:1]:
```suggestion
// Copyright 2021 Tokamak contributors
```

**Comment by MaxDesiatov** [comment]
I've created #365 as a first step towards a proper benchmarking suite. After it's reviewed and merged, we can measure how different approaches to `navigationTitle` and preferences would perform (probably in separate PRs so that this one isn't stuck).

**Comment by foscomputerservices** [comment]
Might I inquire as to the latest status of this pull request?

File: Sources/TokamakCore/Styles/TextFieldStyle.swift
Comment (MaxDesiatov) [inline Sources/TokamakCore/Styles/TextFieldStyle.swift:22]: @carson-katri is this `_AnyTextFieldStyle` refactoring somehow required for the toolbar stuff to work? If not, what's its purpose? Could we maybe move it to a separate PR?

Comment (carson-katri) [reply Sources/TokamakCore/Styles/TextFieldStyle.swift:22]: A `ToolbarTextFieldStyle` is defined/used here: `Sources/TokamakStaticHTML/Views/Navigation/NavigationView.swift`

**Comment by MaxDesiatov** [comment]
@carson-katri feel free to pick it up again if you have a moment. Or let me know otherwise, I hope I can get it over the line in the next few weeks.

**Review by MaxDesiatov — APPROVED** [review]

**Pull request merged**
